src/components/intent.ts

```diff
@@ -0,0 +1,850 @@
+        registered?: boolean;
+    }
+    // These two are only used if no opts.backingStore is provided to the constructor.
+    private readonly _membershipStates: Record<string, MembershipState> = {};
+    private readonly _powerLevels: Record<string, PowerLevelContent> = {};
+
+    /**
+    * Create an entity which can fulfil the intent of a given user.
+    * @constructor
+    * @param client The matrix client instance whose intent is being
+    * fulfilled e.g. the entity joining the room when you call intent.join(roomId).
+    * @param botClient The client instance for the AS bot itself.
+    * This will be used to perform more priveleged actions such as creating new
+    * rooms, sending invites, etc.
+    * @param opts Options for this Intent instance.
+    * @param opts.registered True to inform this instance that the client
+    * is already registered. No registration requests will be made from this Intent.
+    * Default: false.
+    * @param opts.dontCheckPowerLevel True to not check for the right power
+    * level before sending events. Default: false.
+    *
+    * @param opts.backingStore An object with 4 functions, outlined below.
+    * If this Object is supplied, ALL 4 functions must be supplied. If this Object
+    * is not supplied, the Intent will maintain its own backing store for membership
+    * and power levels, which may scale badly for lots of users.
+    *
+    * @param opts.backingStore.getMembership A function which is called with a
+    * room ID and user ID which should return the membership status of this user as
+    * a string e.g "join". `null` should be returned if the membership is unknown.
+    *
+    * @param opts.backingStore.getPowerLevelContent A function which is called
+    * with a room ID which should return the power level content for this room, as an Object.
+    * `null` should be returned if there is no known content.
+    *
+    * @param opts.backingStore.setMembership A function with the signature:
+    * function(roomId, userId, membership) which will set the membership of the given user in
+    * the given room. This has no return value.
+    *
+    * @param opts.backingStore.setPowerLevelContent A function with the signature:
+    * function(roomId, content) which will set the power level content in the given room.
+    * This has no return value.
+    *
+    * @param opts.dontJoin True to not attempt to join a room before
+    * sending messages into it. The surrounding code will have to ensure the correct
+    * membership state itself in this case. Default: false.
+    *
+    * @param opts.enablePresence True to send presence, false to no-op.
+    *
+    * @param opts.caching.ttl How long requests can stay in the cache, in milliseconds.
+    * @param opts.caching.size How many entries should be kept in the cache, before the oldest is dropped.
+    */
+    constructor(private client: MatrixClient, private botClient: MatrixClient, opts: IntentOpts = {}) {
```

Comment:
client shouldn't be private, it's deliberately exposed for bridges to use if they need access to the js-sdk